File: js/analysis/cpp/src/Interface/base256.cpp
#include "base256.h"
#include "errors.h"

void ui_b256(vector<uint8_t>& vec, int num, size_t chars, bool atEnd) {
  int i = num;
  if (i < 0) {
    fatalError(
      spf("ui_b256 can convert to Base256 only non-negative numbers, but got %d", i));
  }
  for (int c = 0; c < chars; ++c) {
    if (atEnd) {
      vec.insert(vec.end() - c, i % 256);
    }
    else {
      vec.insert(vec.begin(), i % 256);
    }
    i /= 256;
  }
  if (i > 0) {
    fatalError(
      spf("ui_b256 cannot convert number %d to %lu base256 characters due to overflow",
        num, chars));
  }
}

void safeString_b256(vector<uint8_t>& vec, string st, size_t sizeCharacters) {
  ui_b256(vec, st.size(), sizeCharacters);
  for (int i=0; i<st.size(); ++i) {
    vec.push_back(st[i]);
  }
}

int b256_ui(const uint8_t* buf, size_t& pos, size_t chars) {
  int res = 0;
  int pow = 1;
  for (size_t c = 0; c < chars; ++c) {
    int data = buf[pos + chars - 1 - c];
    res += data * pow;
    pow *= 256;
  }
  pos += chars;
  //printf("b256_ui %zu %zu %d\n", pos - chars, chars, res);
  return res;
}

string b256_safeString(const uint8_t *buf, size_t &pos, size_t sizeCharacters) {
  int chars = b256_ui(buf, pos, sizeCharacters);
  string res(buf + pos, buf + pos + sizeCharacters + chars);
  pos += chars;
  return res;
}

uint8_t* vector2byteArray(vector<uint8_t>& vec) {
  int sizeChars = log(vec.size()) / log(256) + 1;
  ui_b256(vec, vec.size(), sizeChars, false);
  uint8_t* result = new uint8_t[vec.size() + 1];
  result[0] = sizeChars;
  for (int i=0; i<vec.size(); ++i) {
    result[i + 1] = vec[i];
  }
  return result;
}

File: docs/prod/js/core/base256.js
export function ui_b256(num, chars=1) {
  if (num == null) {
    console.trace();
    throw "Trying to serialize null";
  }
  //console.log('ui_b256', num);
  //num = num || 0;
  let i = Math.floor(num);
  if (i < 0) {
    throw("ui_b256 can convert to Base256 only non-negative numbers, but got " + i);
  }
  let st = '';
  for (let c = 0; c < chars; ++c) {
    let data = i % 256;
    //console.log('Ascii', data, String.fromCharCode(data), encodeURI(String.fromCharCode(data)), String.fromCharCode(data).charCodeAt(0));
    st = String.fromCharCode(data) + st;
    i = Math.floor(i / 256);
  }
  if (i > 0) {
    throw(`ui_b256 cannot convert number ${num} to ${chars} base256 characters due to overflow`);
  }
  //console.log('ui_b64', num, chars, st);
  return st;
}

export function b256_single(char) {
  return char.charCodeAt(0);
}

export function b256_ui(st, pos, chars=1) {
  let res = 0;
  let pow = 1;
  for (let c = 0; c < chars; ++c) {
    let data = b256_single(st[pos[0] + chars - 1 - c]);
    res += data * pow;
    pow *= 256;
  }
  pos[0] += chars;
  //console.log('b256_ui', pos[0] - chars, chars, st.substr(pos[0] - chars, chars), res);
  return res;
}

export function safeString_b256(st, sizeCharacters=1) {
  st = st ? st : "";
  return ui_b256(st.length, sizeCharacters) + st;
}

export function b256_safeString(st, pos, sizeCharacters=1) {
  let chars = b256_ui(st, pos, sizeCharacters);
  let res = st.substr(pos[0], chars);
  //console.log('b64_safeShortString', chars, res);
  pos[0] += chars;
  return res;
}

export function b256_debugSingle(char) {
  let res = '';
  let data = b256_single(char);
  res += data;
  if (data > 32 && data < 126) {
    res += ' ' + char;
  }
  res += '~';
  return res;
}

export function b256_debug(st) {
  let res = '';
  for (let i=0; i<st.length; ++i) {
    res += i + '#';
    res += b256_debugSingle(st[i]);
  }
  return res;
}